### PR TITLE
vspec2graphql.py: (New GraphQL schema generator)

### PR DESCRIPTION
**Pull request by gunnarx**
Based on the the protobuf generator and the output according to the "vehicle.ts" file that was previously provided for the CCS-arch-demo (which we only had the file for, not the generator that created it!)

### CODE REVIEW

File: contrib/vspec2graphql.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        graphql_file.write(indent + f"type {tree_node.qualified_name('_')} {{" + "\n")
+        print_message_body(tree_node.children, graphql_file)
+        graphql_file.write(indent + "}\n\n")
+
+def print_message_body(nodes, graphql_file):
+    for i, node in enumerate(nodes, 1):
+        data_type = node.qualified_name('_')
+        if not node.is_branch():
+            dt_val = node.data_type.value
+            # TODO: FIX ARRAYS!
```
Comment (erikbosch) [inline contrib/vspec2graphql.py:64]: What is it that is left to do?

Comment (gunnarx) [reply contrib/vspec2graphql.py:64]: Good catch.  I already added array support, so this comment should be removed.

Comment (gunnarx) [reply contrib/vspec2graphql.py]: New version below.  I rewrote a few comments for clarity.

File: contrib/vspec2graphql.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        # CamelCase but starting lower case
+        name = node.name[0].lower() + node.name[1:]
+        graphql_file.write(indent * 2 + '"""\n')
+        graphql_file.write(indent * 2 + node.description + "\n")
+        graphql_file.write(indent * 2 + '"""\n')
+        graphql_file.write(indent * 2 + f"{name}: {data_type}\n\n")
+
+def usage():
+    print(
+        """Usage: vspec2graphqlbuf.py [-I include_dir] ... [-i prefix:id_file] vspec_file output_file
```
Comment (erikbosch) [inline contrib/vspec2graphql.py:80]: Not "vspec2graphqlbuf.py" right? Remove "buf"?

File: contrib/vspec2graphql.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  -I include_dir       Add include directory to search for included vspec
+                       files. Can be used multiple times.
+
+  -i prefix:uuid_file  File to use for storing generated UUIDs for signals with
```
Comment (erikbosch) [inline contrib/vspec2graphql.py:84]: called id_file on line 80, uuid_file here. Use same?

File: contrib/vspec2graphql.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                       a given path prefix. Can be used multiple times to store
+                       UUIDs for signal sub-trees in different files.
+
+ vspec_file            The vehicle specification file to parse.
+ graphql_file            The file to output the graphql file to.)
```
Comment (erikbosch) [inline contrib/vspec2graphql.py:89]: Called output_file on oline 80, graphql_file here, align?

Comment (gunnarx) [reply contrib/vspec2graphql.py:89]: Fixed

File: contrib/vspec2graphql.py (hunk 3 above)
Comment (gunnarx) [reply contrib/vspec2graphql.py:84]: Aligned. 

I'm wondering if a tool like this should trigger a rewrite of the UUID file or not.  The others do, so this one also does, but I'm not personally so invested in the UUID concept.

File: contrib/vspec2graphql.py (hunk 2 above)
Comment (gunnarx) [reply contrib/vspec2graphql.py:80]: Fixed

File: contrib/vspec2graphql.py (hunk 3 above)
Comment (erikbosch) [reply contrib/vspec2graphql.py:84]: Do we have any documentation somewhere on how uuid and prefix are supposed to be used? Is it someone that actually use them and the generated uuid_file today (Could be a topic for general tool documentation)

**Comment by gunnarx** [comment]
I added a check to not generate types for empty branches, but that is not enough because a branch could have a sub-branch but the sub-branch is empty.  So it needs a better solution:

For example.

1. Recursive/iterative traversal down the branch to see if there exists _any_ non-branch children
2. or add a file that specifies a filter for what to include in the generation (so that problematic branches in current VSS master can be avoided)

(I think the empty branches should be considered as a bug in the VSS catalog but that doesn't help the issue at the moment)

**Comment by gunnarx** [comment]
Merging, as agreed in https://github.com/GENIVI/vehicle_signal_specification/pull/290.  A better empty-branch check is still pending but we might just avoid having empty branches instead.  We will see.

**Pull request merged**